Apps/Src/CommandLine.c

```diff
@@ -0,0 +1,357 @@
+		printf("> ");
+		BSP_UART_Read(UART_2, input);
+		printf("\n\r");
+
+		if (!executeCommand(input)) { // If command failed, error
+			printf("Bad cmd. Please try again\n\r");
+		}
+	}
+
+    // Delay of 1 seconds
+    OSTimeDlyHMSM(0, 0, 1, 0, OS_OPT_TIME_HMSM_STRICT, &err);
+    if (err != OS_ERR_NONE){
+        assertOSError(OS_NONE_LOC, err);
+    }
+}
+
+// Function Implementations
+// ------------------------------------------------------------------------
+
+static inline bool cmd_help(void) {
+	printf(help);
+	return true;
+}
+
+/* This has not been tested and not sure if this implementation is correct (not sure
+if we want to just transmit string data from serial) */
+static bool cmd_CANbus_Send(void){
+	char *data = strtok_r(NULL, " ", &save);
+	CANDATA_t msg = {.ID=0x0582, .idx=0};		// this would change in the future (don't assume char as data)
+	for(int i = 0; i < 8 && i < strlen(data); i++){
+		msg.data[i] = data[i];
+	}
+
+	char *blockInput = strtok_r(NULL, " ", &save);
+	bool blocking;
+	if(strcmp(blockInput, "blocking") == 0){
+		blocking = CAN_BLOCKING;
+	}
+	else if(strcmp(blockInput, "nonblocking") == 0){
+		blocking = CAN_NON_BLOCKING;
+	}
+	else{
+		return false;
+	}
+
+	char *busInput = strtok_r(NULL, " ", &save);
+	CAN_t bus;
+	if(strcmp(busInput, "motor") == 0){
+		bus = CAN_1;
+	}
+	else if(strcmp(busInput, "car") == 0){
+		bus = CAN_3;
+	}
+	else{
+		return false;
+	}
+
+	if(CANbus_Send(msg, blocking, bus)){
+		printf("msg sent on %s (%s)\n\r", "can", blockInput);
+	}else{
+		printf("msg sent failed\n\r");
+	}
+	return true;
+}
+
+static bool cmd_CANbus_Read(void){
+	CANDATA_t msg;
+
+	char *blockInput = strtok_r(NULL, " ", &save);
+	bool blocking;
+	if(strcmp(blockInput, "blocking") == 0){
+		blocking = CAN_BLOCKING;
+	}
+	else if(strcmp(blockInput, "nonblocking") == 0){
+		blocking = CAN_NON_BLOCKING;
+	}
+	else{
+		return false;
+	}
+
+	char *busInput = strtok_r(NULL, " ", &save);
+	CAN_t bus;
+	if(strcmp(busInput, "motor") == 0){
+		bus = CAN_1;
+	}
+	else if(strcmp(busInput, "car") == 0){
+		bus = CAN_3;
+	}
+	else{
+		return false;
+	}
+
+	if(CANbus_Read(&msg, blocking, bus) == SUCCESS){
+		printf("msg recieved on %s (%s)\n\r", busInput, blockInput);
+		printf("ID: %d, Data: ", msg.ID);
+		for(int i = 0; i < 8; i++){
+			printf("[%d] %x \n\r", i, msg.data[i]);
+		}
+	}else{
+		printf("read failed on %s (%s)\n\r", busInput, blockInput);
+	}
+	return true;
+}
+
+static bool cmd_Contactors_Get(void){
+	char *contactorInput = strtok_r(NULL, " ", &save);
+	contactor_t contactor;
+	if(strcmp(contactorInput, "array_c") == 0){
+		contactor = ARRAY_CONTACTOR;
+	}
+	else if(strcmp(contactorInput, "array_p") == 0){
+		contactor = ARRAY_PRECHARGE;
+	}
+	else if(strcmp(contactorInput, "motor_c") == 0){
+		contactor = MOTOR_CONTACTOR;
+	}
+	else{
+		return false;
+	}
+
+	printf("%s state: %s\n", contactorInput, Contactors_Get(contactor) == ON ? "on" : "off");
+	return true;
+}
+
+static bool cmd_Contactors_Set(void){
```

Comment:
Might run into competition against other threads / logic when this is being used to manually set a contactor. Be interested to see what happens in real life / testing.

Reply:
I don't think it would matter since the contactors is protected by a mutex, so each thread would wait. But we can confirm this in testing. @Cam0Cow

Reply:
I think this one would make sense, but I've yet to test it